src/igraph_eulerian.c

```diff
@@ -0,0 +1,652 @@
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    curr = start_of_path;
+
+    while (!igraph_stack_empty(&tracker)) {
+        
+        if (VECTOR(degree)[curr] != 0) {
+            igraph_vector_int_t *incedges;
+            long nc, edge;
+            long int j, next;
+            IGRAPH_CHECK(igraph_stack_push(&tracker, curr));
+            
+            incedges = igraph_inclist_get(&il, curr);
+            nc = igraph_vector_int_size(incedges);
+
+            for (j = 0; j < nc; j++) {
+                edge = (long) VECTOR(*incedges)[j];
+                if (!VECTOR(visited_list)[edge]) {
+                    break;
+                }
+            }
+            
+            next = IGRAPH_OTHER(graph, edge, curr);
+
+            IGRAPH_CHECK(igraph_stack_push(&edge_tracker, edge));
+
+            /* remove edge here */
+            VECTOR(degree)[curr]--;
+            VECTOR(degree)[next]--;
+            VECTOR(visited_list)[edge] = 1;
+
+            curr = next;
+        } else { /* back track to find remaining circuit */
+            igraph_integer_t curr_e;
+            IGRAPH_CHECK(igraph_stack_push(&path, curr));
+            curr = igraph_stack_pop(&tracker);
+            if (!igraph_stack_empty(&edge_tracker)) {
+                curr_e = igraph_stack_pop(&edge_tracker);
+                IGRAPH_CHECK(igraph_stack_push(&edge_path, curr_e));
+            }
+        }
+    }
+
+    if (edge_res) {
+        while (!igraph_stack_empty(&edge_path)) {
+            IGRAPH_CHECK(igraph_vector_push_back(edge_res, igraph_stack_pop(&edge_path)));           
+        }
+    }
+    if (vector_res) {
+        while (!igraph_stack_empty(&path)) {
+            IGRAPH_CHECK(igraph_vector_push_back(vector_res, igraph_stack_pop(&path)));           
+        }
+    }
+
+    igraph_stack_destroy(&path);
+    igraph_stack_destroy(&tracker);
+    igraph_stack_destroy(&edge_path);
+    igraph_stack_destroy(&edge_tracker);
+    igraph_vector_bool_destroy(&visited_list);
+    igraph_inclist_destroy(&il);
+    igraph_vector_destroy(&degree);    
+    IGRAPH_FINALLY_CLEAN(7);
+
+    return IGRAPH_SUCCESS;
+}
+
+/* solution adapted from https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/ */
+static int igraph_i_eulerian_path_directed(const igraph_t *graph, igraph_vector_t *edge_res, igraph_vector_t *vector_res, igraph_integer_t start_of_path) {
+    long int curr;
+    igraph_integer_t n, m;
+    igraph_inclist_t il;
+    igraph_stack_t path, tracker, edge_tracker, edge_path;
+    igraph_vector_bool_t visited_list;
+    igraph_vector_t remaining_out_edges;
+
+    n = igraph_vcount(graph);
+    m = igraph_ecount(graph);
+
+    if (edge_res) {
+        igraph_vector_clear(edge_res);
+        IGRAPH_CHECK(igraph_vector_reserve(edge_res, m));
+    }
+
+    if (vector_res) {
+        igraph_vector_clear(vector_res);
+        IGRAPH_CHECK(igraph_vector_reserve(vector_res, n+1));
```

Comment:
Same as above, this should be `m+1`, not `n+1`.

Reply:
Addressed in 2d36c3c